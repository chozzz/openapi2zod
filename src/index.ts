export * from './openapi2zod';