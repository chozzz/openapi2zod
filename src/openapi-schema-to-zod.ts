import { z, ZodEffects, ZodString, ZodTypeAny, ZodObject, ZodDefault } from 'zod'
import { OpenAPIV3 } from 'openapi-types'
import { flattenZodType } from './utils/flattenZodType'
import { applyDefaultValue } from './utils/applyDefaultValue'
import zodToJsonSchema from 'zod-to-json-schema';

/**
 * Recursively collects default values from an object schema.
 * @param schema - The OpenAPI schema object.
 * @returns - A partial object containing only default values.
 */
function getDefaultValues(schema: OpenAPIV3.SchemaObject): Record<string, any> | undefined {
  if (!schema || typeof schema !== 'object' || schema.type !== 'object' || !schema.properties) {
    return undefined;
  }

  const defaultValues: Record<string, any> = {};
  let hasDefaults = false;

  for (const [key, propertySchema] of Object.entries(schema.properties)) {
    const propertyObject = propertySchema as OpenAPIV3.SchemaObject;

    // Check if this property has a default value
    if (propertyObject.default !== undefined) {
      defaultValues[key] = propertyObject.default;
      hasDefaults = true;
    }

    // Recursively check if nested objects have defaults
    if (propertyObject.type === 'object' && propertyObject.properties) {
      const nestedDefaults = getDefaultValues(propertyObject);
      if (nestedDefaults !== undefined) {
        // Merge the nested default values
        defaultValues[key] = { ...defaultValues[key], ...nestedDefaults };
        hasDefaults = true;
      }
    }
  }

  // Return the object with default values if any were found, else undefined
  return hasDefaults ? defaultValues : undefined;
}

/**
 * Converts an OpenAPI schema object to a Zod schema.
 * @param schema - OpenAPI SchemaObject or ReferenceObject
 * @param schemas - Map of component schemas for resolving $ref
 * @returns - ZodTypeAny
 */
export function openApiSchemaToZod(
  schema: OpenAPIV3.SchemaObject | OpenAPIV3.ReferenceObject | undefined,
  schemas: Record<string, OpenAPIV3.SchemaObject>,
): ZodTypeAny {
  if (!schema || typeof schema !== 'object') {
    return z.any()
  }

  // Handle $ref references
  if ('$ref' in schema && schema.$ref) {
    const refPath = schema.$ref.replace(/^#\/components\/schemas\//, '')
    const refSchema = schemas[refPath]
    if (refSchema) {
      return openApiSchemaToZod(refSchema, schemas)
    } else {
      console.warn(`Reference ${schema.$ref} not found in schemas.`)
      return z.any()
    }
  }

  // Handle composite schemas
  if ('oneOf' in schema && schema.oneOf) {
    const zodSchemas = schema.oneOf.map((s) => openApiSchemaToZod(s, schemas))

    let combinedSchema: ZodTypeAny

    if (zodSchemas.length === 0) {
      combinedSchema = z.any()
    } else if (zodSchemas.length === 1) {
      combinedSchema = zodSchemas[0]
    } else {
      combinedSchema = z.union(zodSchemas as [ZodTypeAny, ZodTypeAny, ...ZodTypeAny[]])
    }

    // Attach description if available
    if ('description' in schema && schema.description) {
      combinedSchema = combinedSchema.describe(schema.description)
    }

    return combinedSchema
  }

  if ('anyOf' in schema && schema.anyOf) {
    const zodSchemas = schema.anyOf.map((s) => openApiSchemaToZod(s, schemas))

    let combinedSchema: ZodTypeAny

    if (zodSchemas.length === 0) {
      combinedSchema = z.any()
    } else if (zodSchemas.length === 1) {
      combinedSchema = zodSchemas[0]
    } else {
      combinedSchema = z.union(zodSchemas as [ZodTypeAny, ZodTypeAny, ...ZodTypeAny[]])
    }

    // Attach description if available
    if ('description' in schema && schema.description) {
      combinedSchema = combinedSchema.describe(schema.description)
    }

    return combinedSchema
  }

  if ('allOf' in schema && schema.allOf) {
    const zodSchemas = schema.allOf.map((s) => openApiSchemaToZod(s, schemas))

    let combinedSchema: ZodTypeAny

    if (zodSchemas.length === 0) {
      combinedSchema = z.any()
    } else {
      combinedSchema = zodSchemas.reduce((prev, curr) => z.intersection(prev, curr))
    }

    // **Flatten the intersection of ZodObjects**
    combinedSchema = flattenZodType(combinedSchema)

    // Attach description if available
    if ('description' in schema && schema.description) {
      combinedSchema = combinedSchema.describe(schema.description)
    }

    return combinedSchema
  }

  if ('not' in schema && schema.not) {
    let combinedSchema = z
      .any()
      .refine((value) => !openApiSchemaToZod(schema.not, schemas).safeParse(value).success)

    // Attach description if available
    if ('description' in schema && schema.description) {
      combinedSchema = combinedSchema.describe(schema.description)
    }

    return combinedSchema
  }

  if ('type' in schema) {
    switch (schema.type) {
      case 'string': {
        // Handle enums separately
        if ('enum' in schema && schema.enum) {
          const enumValues = schema.enum as [string, ...string[]]
          let enumSchema = z.enum(enumValues)

          // Attach description if available
          if ('description' in schema && schema.description) {
            enumSchema = enumSchema.describe(schema.description)
          }

          return enumSchema
        }

        // Start with a ZodString
        let stringSchema: ZodString | ZodEffects<ZodString, any, any> | ZodDefault<ZodString> = z.string()

        // Handle formats
        if ('format' in schema && schema.format) {
          switch (schema.format) {
            case 'email':
              stringSchema = stringSchema.email()
              break
            case 'uuid':
              stringSchema = stringSchema.uuid()
              break
            case 'uri':
              stringSchema = stringSchema.url(); // Strict URL validation for 'uri'
              break;
            case 'uri-reference':
              // Custom validation for 'uri-reference' to allow paths and fragments
              stringSchema = stringSchema.refine(
                (val) => typeof val === 'string' && /^[a-zA-Z0-9\/#\?&\.]+$/.test(val),
                { message: 'Invalid uri-reference format' }
              );
              break;
            case 'hostname':
              stringSchema = stringSchema.refine(
                (val) => {
                  // Simple regex for hostname validation
                  const hostnameRegex =
                    /^(?=.{1,253}$)(?!\-)(?:[a-zA-Z0-9\-]{0,62}[a-zA-Z0-9]\.)+[a-zA-Z]{2,63}$/
                  return hostnameRegex.test(val)
                },
                { message: 'Invalid hostname format' },
              )
              break
            case 'ipv4':
              stringSchema = stringSchema.ip({ version: 'v4' })
              break
            case 'ipv6':
              stringSchema = stringSchema.ip({ version: 'v6' })
              break
            case 'date-time':
              stringSchema = stringSchema.datetime({ offset: true })
              break
            case 'date':
              stringSchema = stringSchema.refine(
                (val) => {
                  // Simple date validation (YYYY-MM-DD)
                  const dateRegex = /^\d{4}-\d{2}-\d{2}$/
                  return dateRegex.test(val)
                },
                { message: 'Invalid date format' },
              )
              break
            case 'time':
              stringSchema = stringSchema.refine(
                (val) => {
                  // Simple time validation (HH:MM:SS)
                  const timeRegex = /^([01]\d|2[0-3]):[0-5]\d:[0-5]\d$/
                  return timeRegex.test(val)
                },
                { message: 'Invalid time format' },
              )
              break
            case 'byte':
              stringSchema = stringSchema.base64()
              break
            case 'password':
              // For passwords, we might not enforce any format, but you can add custom rules
              // For example, minimum length
              stringSchema = stringSchema.min(8, 'Password must be at least 8 characters long')
              break
            case 'binary':
              // Handle binary data as Buffer instance
              return z.instanceof(Buffer)
            default:
              console.warn(`Unhandled string format: ${schema.format}`)
          }
        }

        // Attach description if available
        if ('description' in schema && schema.description) {
          stringSchema = stringSchema.describe(schema.description)
        }

        // Apply default value if available
        if ('default' in schema && schema.default) {
          stringSchema = applyDefaultValue(stringSchema as ZodString, schema.default)
        }

        return stringSchema
      }
      case 'integer': {
        let integerSchema = z.number().int()

        // Attach description if available
        if ('description' in schema && schema.description) {
          integerSchema = integerSchema.describe(schema.description)
        }

        // Apply default value if available
        if ('default' in schema && schema.default) {
          integerSchema = applyDefaultValue(integerSchema, schema.default) as z.ZodNumber;
        }

        return integerSchema
      }
      case 'number': {
        let numberSchema = z.number()

        // Attach description if available
        if ('description' in schema && schema.description) {
          numberSchema = numberSchema.describe(schema.description)
        }

        // Apply default value if available
        if ('default' in schema && schema.default) {
          numberSchema = applyDefaultValue(numberSchema, schema.default) as z.ZodNumber;
        }

        return numberSchema
      }
      case 'boolean': {
        let booleanSchema = z.boolean()

        // Attach description if available
        if ('description' in schema && schema.description) {
          booleanSchema = booleanSchema.describe(schema.description)
        }

        // Apply default value if available
        if ('default' in schema && schema.default) {
          booleanSchema = applyDefaultValue(booleanSchema, schema.default) as z.ZodBoolean;
        }

        return booleanSchema
      }
      case 'array': {
        if ('items' in schema && schema.items) {
          let arraySchema = z.array(openApiSchemaToZod(schema.items, schemas))

          // Attach description if available
          if ('description' in schema && schema.description) {
            arraySchema = arraySchema.describe(schema.description)
          }

          // Apply default value if available
          if ('default' in schema && schema.default) {
            arraySchema = applyDefaultValue(arraySchema, schema.default) as z.ZodArray<ZodTypeAny, "many">;
          }

          return arraySchema
        } else {
          console.warn("Array schema without 'items' property.")
          return z.array(z.any())
        }
      }
      case 'object': {
        const properties = schema.properties || {};
        const requiredProps = schema.required || [];
        const zodProperties: Record<string, ZodTypeAny> = {};
    
        for (const [key, value] of Object.entries(properties)) {
          let propertySchema = openApiSchemaToZod(value as OpenAPIV3.SchemaObject, schemas);
    
          if (!requiredProps.includes(key)) {
            propertySchema = propertySchema.optional();
          }
    
          if ('description' in value && value.description) {
            propertySchema = propertySchema.describe(value.description);
          }
    
          if ('default' in value && value.default !== undefined) {
            propertySchema = applyDefaultValue(propertySchema, value.default);
          }
    
          zodProperties[key] = propertySchema;
        }
    
        let objectSchema: ZodObject<Record<string, ZodTypeAny>> = z.object(zodProperties);
    
        if (schema.additionalProperties) {
          if (schema.additionalProperties === true) {
            objectSchema = objectSchema.catchall(z.any());
          } else if (typeof schema.additionalProperties === 'object') {
            objectSchema = objectSchema.catchall(openApiSchemaToZod(schema.additionalProperties, schemas));
          }
        }
    
        if ('description' in schema && schema.description) {
          objectSchema = objectSchema.describe(schema.description);
        }
    
        // Set parent default value based on child defaults
        const defaultValues = getDefaultValues(schema);

        if (defaultValues !== undefined) {
          objectSchema = objectSchema.default(defaultValues) as unknown as ZodObject<Record<string, ZodTypeAny>, "strip", ZodTypeAny, { [x: string]: any }, { [x: string]: any }>;
        }
        
        // Use transform to enforce default structure when parsing
        objectSchema = objectSchema.transform((input, ctx) => {
          // Merge defaults with input, ensuring nested defaults are respected
          return { ...defaultValues, ...input };
        }) as unknown as ZodObject<Record<string, ZodTypeAny>, "strip", ZodTypeAny, { [x: string]: any }, { [x: string]: any }>;

    
        return objectSchema;
      }
      default:
        console.warn(`Unhandled schema type: ${schema.type}`)
        return z.any()
    }
  } else {
    console.warn("Schema without 'type' property.")
    return z.any()
  }
}
