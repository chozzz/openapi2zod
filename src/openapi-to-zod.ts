import { z, ZodSchema, ZodTypeAny } from "zod";
import { OpenAPIV3 } from "openapi-types";
import { openApiSchemaToZod } from "./openapi-schema-to-zod";

/**
 * Parses an OpenAPI schema and returns Zod schemas for all operations.
 * @param openApiSchema - The OpenAPI schema object
 * @returns - A record mapping operation IDs to their Zod schemas
 */
export function parseOpenApiToZod(openApiSchema: OpenAPIV3.Document): Record<string, ZodSchema> {
  const zodSchemas: Record<string, ZodSchema> = {};

  const paths = openApiSchema.paths || {};
  const schemas = openApiSchema.components?.schemas || {};

  // Convert schemas to proper mapping for reference resolution
  const schemaMap: Record<string, OpenAPIV3.SchemaObject> = {};
  for (const [schemaName, schemaDef] of Object.entries(schemas)) {
    if (!('$ref' in schemaDef)) {
      schemaMap[schemaName] = schemaDef;
    }
  }

  for (const [path, pathItem] of Object.entries(paths)) {
    if (!pathItem) continue;

    for (const method of ['get', 'post', 'put', 'patch', 'delete', 'options', 'head', 'trace'] as const) {
      const operation = pathItem[method] as OpenAPIV3.OperationObject | undefined;
      if (!operation) continue;

      const operationId = operation.operationId || `${method.toUpperCase()} ${path}`;

      const parameters = [
        ...(pathItem.parameters || []),
        ...(operation.parameters || []),
      ] as OpenAPIV3.ParameterObject[];

      const requestBody = operation.requestBody;

      // Collect all parameters and request body properties
      const combinedSchemas: Record<string, ZodTypeAny> = {};

      // Process parameters
      for (const param of parameters) {
        const { name, required } = param;
        let paramSchema: ZodTypeAny;

        if ('schema' in param && param.schema) {
          paramSchema = openApiSchemaToZod(param.schema, schemaMap);
        } else {
          console.warn(`Parameter ${name} missing schema.`);
          paramSchema = z.any();
        }
        
        // Attach description to object if available
        if (param?.description && !paramSchema.description) {
          paramSchema = paramSchema.describe(param.description);
        }

        if (!required) {
          paramSchema = paramSchema.optional();
        }

        combinedSchemas[name] = paramSchema;
      }

      // Process requestBody
      let requestBodySchema: ZodTypeAny | undefined = undefined;

      if (requestBody && 'content' in requestBody && requestBody.content) {
        const requestBodyObj = requestBody as OpenAPIV3.RequestBodyObject;
        const jsonContent = requestBodyObj.content["application/json"];
        if (jsonContent && 'schema' in jsonContent && jsonContent.schema) {
          requestBodySchema = openApiSchemaToZod(jsonContent.schema, schemaMap);

          if (requestBodySchema instanceof z.ZodObject) {
            Object.assign(combinedSchemas, requestBodySchema.shape);
          } else {
            combinedSchemas['requestBody'] = requestBodySchema;
          }
        } else {
          console.warn(`Request body for operation ${operationId} missing 'application/json' content schema.`);
        }
      }

      // Create the operation schema
      const operationSchema = z.object(combinedSchemas);

      // Add the schema to the result
      zodSchemas[operationId] = operationSchema;
    }
  }

  return zodSchemas;
}
