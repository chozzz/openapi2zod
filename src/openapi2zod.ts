import { z, ZodSchema, ZodTypeAny } from 'zod'
import { OpenAPIV3 } from 'openapi-types'
import { openApiSchemaToZod } from './openapi-schema-to-zod'
import { flattenZodType } from './utils/flattenZodType'

export function parseOpenApiToZod(openApiSchema: OpenAPIV3.Document): Record<string, ZodSchema> {
  const zodSchemas: Record<string, ZodSchema> = {}

  const paths = openApiSchema.paths || {}
  const schemas = openApiSchema.components?.schemas || {}

  // Convert schemas to proper mapping for reference resolution
  const schemaMap: Record<string, OpenAPIV3.SchemaObject> = {}
  for (const [schemaName, schemaDef] of Object.entries(schemas)) {
    if (!('$ref' in schemaDef)) {
      schemaMap[schemaName] = schemaDef
    }
  }

  for (const [path, pathItem] of Object.entries(paths)) {
    if (!pathItem) continue

    for (const method of [
      'get',
      'post',
      'put',
      'patch',
      'delete',
      'options',
      'head',
      'trace',
    ] as const) {
      const operation = pathItem[method] as OpenAPIV3.OperationObject | undefined
      if (!operation) continue

      const operationId = operation.operationId || `${method.toUpperCase()} ${path}`

      const parameters = [
        ...(pathItem.parameters || []),
        ...(operation.parameters || []),
      ] as OpenAPIV3.ParameterObject[]

      const requestBody = operation.requestBody;

      // Collect all parameters and request body properties
      const combinedSchemas: Record<string, ZodTypeAny> = {}

      // Process parameters
      for (const param of parameters) {
        const { name, required } = param
        let paramSchema: ZodTypeAny

        if ('schema' in param && param.schema) {
          paramSchema = openApiSchemaToZod(param.schema, schemaMap)
        } else {
          console.warn(`Parameter ${name} missing schema.`)
          paramSchema = z.any()
        }

        // Attach description to object if available
        if (param?.description && !paramSchema.description) {
          paramSchema = paramSchema.describe(param.description)
        }

        if (!required) {
          paramSchema = paramSchema.optional()
        }

        combinedSchemas[name] = paramSchema
      }

      // Create parameters schema
      const parametersSchema = z.object(combinedSchemas)

      // Process requestBody
      let requestBodySchema: ZodTypeAny | undefined = undefined

      if (requestBody && 'content' in requestBody && requestBody.content) {
        const requestBodyObj = requestBody as OpenAPIV3.RequestBodyObject
        const jsonContent = requestBodyObj.content['application/json']
        if (jsonContent && 'schema' in jsonContent && jsonContent.schema) {
          requestBodySchema = openApiSchemaToZod(jsonContent.schema, schemaMap)
        } else {
          console.warn(
            `Request body for operation ${operationId} missing 'application/json' content schema.`,
          )
        }
      }

      let operationSchema: ZodTypeAny;

      if (requestBodySchema) {
        const isParametersSchemaEmpty =
          parametersSchema instanceof z.ZodObject &&
          Object.keys(parametersSchema.shape).length === 0

        if (isParametersSchemaEmpty) {
          // No parameters, use requestBodySchema directly
          operationSchema = requestBodySchema
        } else {
          // Merge parametersSchema and requestBodySchema appropriately
          // Existing logic for merging schemas
          const flattenedRequestBodySchema = flattenZodType(requestBodySchema)

          if (flattenedRequestBodySchema instanceof z.ZodObject) {
            operationSchema = parametersSchema.merge(flattenedRequestBodySchema);
          }
          else if (flattenedRequestBodySchema instanceof z.ZodUnion) {
            // Handle ZodUnion by merging parametersSchema with each option in the union
            const mergedOptions = flattenedRequestBodySchema.options.map((option: z.ZodObject<any> | z.ZodAny) => {
              // Instead of directly merging, combine the fields using intersection or custom logic
              return z.intersection(parametersSchema, option);
            });
          
            if (mergedOptions.length >= 2) {
              operationSchema = z.union(mergedOptions as [ZodTypeAny, ZodTypeAny, ...ZodTypeAny[]]);
            } else if (mergedOptions.length === 1) {
              operationSchema = mergedOptions[0];
            } else {
              operationSchema = parametersSchema;
            }
          }
          
          else {
            // For other types of requestBodySchema, use z.intersection
            operationSchema = z.intersection(parametersSchema, flattenedRequestBodySchema);
          }

          // Apply transformation to ensure merged fields are preserved
          operationSchema = operationSchema.transform((input) => {
            // Merge fields from both schemas
            const paramResult = parametersSchema.safeParse(input).success ? input : {};
            const bodyResult = flattenedRequestBodySchema.safeParse(input).success ? input : {};
            
            return { ...paramResult, ...bodyResult };
          });
        }
      } else {
        operationSchema = parametersSchema
      }

      // Attach description or summary to operationSchema if available
      operationSchema = operationSchema.describe(JSON.stringify({
        description: operation.description || operation.summary || '',
        operationId,
        path,
        method,
      }));

      // Add the schema to the result
      zodSchemas[operationId] = operationSchema
    }
  }

  return zodSchemas
}
