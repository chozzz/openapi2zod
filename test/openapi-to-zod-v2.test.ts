import { OpenAPIV3 } from 'openapi-types'
import { z } from 'zod'
import { parseOpenApiToZod } from '../src/openapi-to-zod' // Adjust the import path accordingly
import { zodToJsonSchema } from 'zod-to-json-schema'
import * as openapiJson from './__mocks__/openapi_3_1_0_v2.json'

describe('parseOpenApiToZod v2', () => {
  let zodSchemas: Record<string, z.ZodSchema>

  beforeAll(() => {
    zodSchemas = parseOpenApiToZod(openapiJson as OpenAPIV3.Document)
  })
  test('should generate correct schema for operationId "oneOfExample"', () => {
    const schema = zodSchemas['oneOfExample'] as z.ZodTypeAny
    expect(schema).toBeDefined()

    // Valid data: conforms to TypeA
    const validDataTypeA = {
      type: 'A',
      value: 'Some value',
    }

    // Valid data: conforms to TypeB
    const validDataTypeB = {
      type: 'B',
      amount: 42,
    }

    // Invalid data: conforms to neither TypeA nor TypeB
    const invalidData = {
      type: 'C',
      value: 'Invalid',
    }

    // Validate parsing
    expect(() => schema.parse(validDataTypeA)).not.toThrow()
    expect(() => schema.parse(validDataTypeB)).not.toThrow()
    expect(() => schema.parse(invalidData)).toThrow()

    // Generate JSON schema and snapshot test
    const jsonSchema = zodToJsonSchema(schema, 'oneOfExample')
    expect(jsonSchema).toMatchSnapshot()
  })
  test('should generate correct schema for operationId "anyOfExample"', () => {
    const schema = zodSchemas['anyOfExample'] as z.ZodTypeAny
    expect(schema).toBeDefined()

    // Valid data: conforms to TypeC
    const validDataTypeC = {
      option1: 'Value',
    }

    // Valid data: conforms to TypeD
    const validDataTypeD = {
      option2: 100,
    }

    // Valid data: conforms to both TypeC and TypeD
    const validDataBoth = {
      option1: 'Value',
      option2: 100,
    }

    // Invalid data: conforms to neither
    const invalidData = {
      option3: 'Invalid',
    }

    // Validate parsing
    expect(() => schema.parse(validDataTypeC)).not.toThrow()
    expect(() => schema.parse(validDataTypeD)).not.toThrow()
    expect(() => schema.parse(validDataBoth)).not.toThrow()
    expect(() => schema.parse(invalidData)).toThrow()

    // Generate JSON schema and snapshot test
    const jsonSchema = zodToJsonSchema(schema, 'anyOfExample')
    expect(jsonSchema).toMatchSnapshot()
  })
  test('should generate correct schema for operationId "allOfExample"', () => {
    const schema = zodSchemas['allOfExample'] as z.ZodTypeAny
    expect(schema).toBeDefined()

    // Valid data: conforms to both BaseType and additional properties
    const validData = {
      id: '123',
      name: 'Test Entity',
      additionalProperty: 'Extra Value',
    }

    // Invalid data: missing property from BaseType
    const invalidDataMissingBase = {
      name: 'Test Entity',
      additionalProperty: 'Extra Value',
    }

    // Invalid data: missing additionalProperty
    const invalidDataMissingAdditional = {
      id: '123',
      name: 'Test Entity',
    }

    // Validate parsing
    expect(() => schema.parse(validData)).not.toThrow()
    expect(() => schema.parse(invalidDataMissingBase)).toThrow()
    expect(() => schema.parse(invalidDataMissingAdditional)).toThrow()

    // Generate JSON schema and snapshot test
    const jsonSchema = zodToJsonSchema(schema, 'allOfExample')
    expect(jsonSchema).toMatchSnapshot()
  })
  test('should generate correct schema for operationId "notExample"', () => {
    const schema = zodSchemas['notExample'] as z.ZodTypeAny
    expect(schema).toBeDefined()

    // Valid data: does not conform to ForbiddenType
    const validData = {
      allowed: true,
    }

    // Invalid data: conforms to ForbiddenType
    const invalidData = {
      forbidden: true,
    }

    // Validate parsing
    expect(() => schema.parse(validData)).not.toThrow()
    expect(() => schema.parse(invalidData)).toThrow()

    // Generate JSON schema and snapshot test
    const jsonSchema = zodToJsonSchema(schema, 'notExample')
    expect(jsonSchema).toMatchSnapshot()
  })
})
