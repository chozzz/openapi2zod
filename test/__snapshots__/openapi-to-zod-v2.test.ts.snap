// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseOpenApiToZod v2 should generate correct schema for operationId "allOfExample" 1`] = `
{
  "$ref": "#/definitions/allOfExample",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "definitions": {
    "allOfExample": {
      "additionalProperties": false,
      "description": "This endpoint accepts a request body that must conform to all of the specified schemas.",
      "properties": {
        "additionalProperty": {
          "description": "An additional property required by this endpoint.",
          "type": "string",
        },
        "id": {
          "description": "Unique identifier",
          "type": "string",
        },
        "name": {
          "description": "Name of the entity",
          "type": "string",
        },
      },
      "required": [
        "id",
        "name",
        "additionalProperty",
      ],
      "type": "object",
    },
  },
}
`;

exports[`parseOpenApiToZod v2 should generate correct schema for operationId "anyOfExample" 1`] = `
{
  "$ref": "#/definitions/anyOfExample",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "definitions": {
    "anyOfExample": {
      "anyOf": [
        {
          "additionalProperties": false,
          "properties": {
            "option1": {
              "type": "string",
            },
          },
          "required": [
            "option1",
          ],
          "type": "object",
        },
        {
          "additionalProperties": false,
          "properties": {
            "option2": {
              "type": "integer",
            },
          },
          "required": [
            "option2",
          ],
          "type": "object",
        },
      ],
      "description": "This endpoint accepts a request body that can be any combination of the specified schemas.",
    },
  },
}
`;

exports[`parseOpenApiToZod v2 should generate correct schema for operationId "notExample" 1`] = `
{
  "$ref": "#/definitions/notExample",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "definitions": {
    "notExample": {
      "description": "This endpoint accepts a request body that must not conform to a specified schema.",
    },
  },
}
`;

exports[`parseOpenApiToZod v2 should generate correct schema for operationId "oneOfExample" 1`] = `
{
  "$ref": "#/definitions/oneOfExample",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "definitions": {
    "oneOfExample": {
      "anyOf": [
        {
          "additionalProperties": false,
          "properties": {
            "type": {
              "enum": [
                "A",
              ],
              "type": "string",
            },
            "value": {
              "type": "string",
            },
          },
          "required": [
            "type",
            "value",
          ],
          "type": "object",
        },
        {
          "additionalProperties": false,
          "properties": {
            "amount": {
              "type": "number",
            },
            "type": {
              "enum": [
                "B",
              ],
              "type": "string",
            },
          },
          "required": [
            "type",
            "amount",
          ],
          "type": "object",
        },
      ],
      "description": "This endpoint accepts a request body that must conform to one of the specified schemas.",
    },
  },
}
`;
